### ISSUE TITLE
Crash with CHECK in check-allocate.cc(199)

### ISSUE THREAD

**Issue by klausler** [flang-compiler/f18#594]
```
MODULE m
  TYPE :: t
    class(t), pointer :: p
   CONTAINS
    PROCEDURE :: f
  END TYPE
 CONTAINS
  FUNCTION f(this) RESULT (x)
    CLASS(t) :: this
    CLASS(t), allocatable :: x
  END FUNCTION
END MODULE

use m
type(t),target :: o
class(t), allocatable :: p
allocate(p, source=o%f())
end
```

**Comment by jeanPerier** [comment]
The `CHECK` is failing after a typeless expression was found in a context where it is not expected. In the bug report case, the internal error is justified because this o%f() has type "t" so something is probably wrong in `GetType()`.
However, investigating further, the `CHECK` is also too strict because it will crash with actual type-less expressions when it should only emit a user error.

I am working on a fix.

**Comment by jeanPerier** [comment]
Fixed with PR #607
